hugr-core/src/export.rs

```diff
@@ -0,0 +1,726 @@
+//! Exporting HUGR graphs to their `hugr-model` representation.
+use crate::{
+    extension::ExtensionSet,
+    ops::OpType,
+    types::{
+        type_param::{TypeArgVariable, TypeParam},
+        type_row::TypeRowBase,
+        CustomType, EdgeKind, FuncTypeBase, MaybeRV, PolyFuncTypeBase, RowVariable, SumType,
+        TypeArg, TypeBase, TypeEnum,
+    },
+    Direction, Hugr, HugrView, IncomingPort, Node, Port, PortIndex,
+};
+use bumpalo::{collections::Vec as BumpVec, Bump};
+use hugr_model::v0::{self as model};
+use indexmap::IndexSet;
+use smol_str::ToSmolStr;
+
+pub(crate) const OP_FUNC_CALL_INDIRECT: &str = "func.call-indirect";
+const TERM_PARAM_TUPLE: &str = "param.tuple";
+
+/// Export a [`Hugr`] graph to its representation in the model.
+pub fn export_hugr<'a>(hugr: &'a Hugr, bump: &'a Bump) -> model::Module<'a> {
+    let mut ctx = Context::new(hugr, bump);
+    ctx.export_root();
+    ctx.module
+}
+
+/// State for converting a HUGR graph to its representation in the model.
+struct Context<'a> {
+    /// The HUGR graph to convert.
+    hugr: &'a Hugr,
+    /// The module that is being built.
+    module: model::Module<'a>,
+    /// Mapping from ports to link indices.
+    /// This only includes the minimum port among groups of linked ports.
+    links: IndexSet<(Node, Port)>,
+    bump: &'a Bump,
+}
+
+impl<'a> Context<'a> {
+    pub fn new(hugr: &'a Hugr, bump: &'a Bump) -> Self {
+        let mut module = model::Module::default();
+        module.nodes.reserve(hugr.node_count());
```

Comment:
Nodes, terms and regions have different capacities, so that'd be a `with_capacity` with quite a few parameters (all of the same type, don't mix them up!).

Reply:
Then maybe a `reserve_nodes` method instead, that can be documented instead of relying on knowing how the Model works internally.

Reply:
There isn't really an internally. It's plain old data. Everything is pub.